tests/benchmark/crud_test.go

```diff
@@ -0,0 +1,920 @@
+	fileBytes, _ := ioutil.ReadFile(filename)
+	var buf bytes.Buffer
+	zw, _ := gzip.NewWriterLevel(&buf, gzip.BestCompression)
+	_, err := zw.Write(fileBytes)
+	if err != nil {
+		return nil, err
+	}
+	if err := zw.Close(); err != nil {
+		return nil, err
+	}
+	return &buf, nil
+}
+func pauseAfter1000Iterations(i int) {
+	if i%1000 == 0 {
+		/*
+			Pause 5 seconds every 1000 iterations to reduce chances of failure on
+			CREATE, UPDATE, DELETE
+		*/
+		time.Sleep(5 * time.Second)
+	}
+}
+
+func BenchmarkCreateApi(b *testing.B) {
+
+	rootResource := getRootResource(b)
+	ctx := context.Background()
+	gapicClient, err := connection.NewClient(ctx)
+	if err != nil {
+		b.Fatalf("Setup: Failed to create client: %s", err)
+	}
+	client := gapicClient.GrpcClient()
+
+	for i := 1; i <= b.N; i++ {
+		pauseAfter1000Iterations(i)
+
+		apiId := getApiName(i, b)
+
+		req := &rpc.CreateApiRequest{
+			Parent: rootResource,
+			ApiId:  apiId,
+			Api: &rpc.Api{
+				DisplayName: apiId,
+				Description: fmt.Sprintf("Description for %s", apiId),
+			},
+		}
+		b.StartTimer()
+		_, err := client.CreateApi(ctx, req)
+		b.StopTimer()
+		/**
+		Go Benchmarking always runs all the benchmark tests once before it runs
+		for the duration specified. So the first row always returns error.
+		*/
+		if i > 1 && err != nil {
+			b.Errorf("CreateApi(%+v) returned unexpected error: %s", req, err)
+		}
+	}
+}
+
+func BenchmarkCreateApiVersion(b *testing.B) {
+	rootResource := getRootResource(b)
+
+	ctx := context.Background()
+	gapicClient, err := connection.NewClient(ctx)
+	if err != nil {
+		b.Fatalf("Setup: Failed to create client: %s", err)
+	}
+	client := gapicClient.GrpcClient()
+
+	b.StopTimer()
+	b.ResetTimer()
+
+	for i := 1; i <= b.N; i++ {
+		pauseAfter1000Iterations(i)
+
+		apiId := getApiName(i, b)
+
+		for j := 1; j <= loadVersionCount(b); j++ {
+			req := &rpc.CreateApiVersionRequest{
+				Parent: fmt.Sprintf("%s/apis/%s", rootResource, apiId),
+				ApiVersion: &rpc.ApiVersion{
+					DisplayName: fmt.Sprintf("v%d", j),
+					Description: fmt.Sprintf("v%d of %s", j, apiId),
+				},
+				ApiVersionId: fmt.Sprintf("v%d", j),
+			}
+			// Only take reading for the first version
+			if j == 1 {
+				b.StartTimer()
+			}
+			_, err := client.CreateApiVersion(ctx, req)
+			if j == 1 {
+				b.StopTimer()
+			}
+			if i > 1 && err != nil {
+				b.Errorf("CreateApiVersion(%+v) returned unexpected error: %s", req, err)
+			}
+		}
+	}
+}
+
+func BenchmarkCreateApiSpec(b *testing.B) {
+	rootResource := getRootResource(b)
+
+	ctx := context.Background()
+	gapicClient, err := connection.NewClient(ctx)
+	if err != nil {
+		b.Fatalf("Setup: Failed to create client: %s", err)
+	}
+	client := gapicClient.GrpcClient()
+
+	b.StopTimer()
+	b.ResetTimer()
+	buf, err := readAndGZipFile(filepath.Join("testdata", "openapi1.yaml"))
+	if err != nil {
+		b.Errorf("BenchmarkCreateApiSpec could not read openapi1.yaml : %s", err)
+	} else {
```

Comment:
This else can be removed. https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

Reply:
You are correct
> If I'm understanding correctly, BenchmarkCreateApiVersion a benchmark and setup code for BenchmarkCreateApiSpec. Is that correct?

I think you are right in terms of we don't need to iterate through every api spec version for the UPDATE/GET calls
That really does not give us any metrics and is wasted execution time. I changed the logic for Update and gets to only operate on the first spec of the first revision of every api.

I left the looping over api versions on the Deletes, to clean everything that was created during the tests.